contracts/CDPLeverage.sol

```diff
@@ -0,0 +1,456 @@
+pragma solidity ^0.4.19;
+
+import "./interfaces/ITub.sol";
+import "./interfaces/DSValue.sol";
+import "./interfaces/IWETH.sol";
+import "./interfaces/IMatchingMarket.sol";
+import "./DSMath.sol";
+
+
+/**
+* CDP-Leverage is a tool to streamline the process of reinvesting in one CDP,
+* allowing you to increase your USD/ETH leverage up to 3x.
+*
+* Please use caution, as the tool is still under development.
+* More information on the tool, visit: https://github.com/KentonPrescott/CDP-leverager
+* More information on Single Collateral Dai, visit: https://github.com/makerdao/sai
+*/
+contract CDPLeverage is DSMath {
+
+    // Constants and modules set at contract inception
+    ITub public tub;
+    DSToken public weth;
+    DSToken public peth;
+    DSToken public dai;
+    DSToken public gov;
+    DSValue public pip;
+    DSValue public pep;
+    uint256 public makerLR;
+    uint256 public fee;
+    uint256 public axe;
+    IMatchingMarket public dex;
+    uint constant WAD = 10 ** 18;
+    uint constant RAY = 10 ** 27;
+
+    // This Struct tracks position information for a specific investor address
+    struct Investor {
+        uint256 layers;          // Number of CDP layers. According to github schematic
+        uint256 principal;       // WAD - Principal contribution in Eth
+        uint256 collatRatio;     // WAD - Collaterazation ratio
+        bytes32 cdpID;           // cdpID
+        uint256 purchPrice;      // WAD; currPrice at time of purchase
+        uint256 daiAmountFinal;  // WAD - Amount of Dai that an investor has after leveraging
+        uint256 totalDebt;       // WAD - Total amount of debt (DAI)
+        uint256 priceFloor;      // WAD - Price floor
+        uint256 index;           // index of investor
+    }
+    mapping (address => Investor) public investors;
+    address[] public investorAddresses;
```

Comment:
I might be missing something w.r.t. how solidity works, but I'm not totally clear on why we need this array

Reply:
Mainly foundational code. The mapping includes every address on the network, and since addresses are not recognizably linked, it is hard to loop through 'investors' that don't carry null/0 fields. 

The array helps to transverse through positions opened via the tool. Though we aren't looping through it a.t.m., it may bring some utility in the future... If need be, we could delete it for gas savings